### PR TITLE
Validate the active bundle

### PR DESCRIPTION
**Pull request by ewollesen**
https://github.com/aws/modelrocket-add-ons/issues/179

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: api/v1alpha1/packagebundlecontroller_webhook.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	return nil
+}
+
+func (v *activeBundleValidator) Handle(ctx context.Context,
+	req admission.Request) admission.Response {
+
+	pbc := &PackageBundleController{}
+	err := v.decoder.Decode(req, pbc)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError,
```
Comment (TerryHowe) [inline api/v1alpha1/packagebundlecontroller_webhook.go:50]: Is this just the first time I noticed the tab stop is 8 here?

File: api/v1alpha1/packagebundlecontroller_webhook.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+			break
+		}
+	}
+
+	resp := &admission.Response{
+		AdmissionResponse: admissionv1.AdmissionResponse{Allowed: found},
+	}
+
+	if !found {
+		msg := fmt.Sprintf("invalid package bundle: %q", pbc.Spec.ActiveBundle)
```
Comment (TerryHowe) [inline api/v1alpha1/packagebundlecontroller_webhook.go:86]:
```suggestion
		msg := fmt.Sprintf("package bundle not found with name: %q", pbc.Spec.ActiveBundle)
```

File: api/v1alpha1/packagebundlecontroller_webhook.go (hunk 1 above)
Comment (ewollesen) [reply api/v1alpha1/packagebundlecontroller_webhook.go:50]: I have no idea. My editor displays each indentation as four spaces regardless of what it really is. I thought gofmt took care of the rest?

I can look into editor settings if you think I have something misconfigured.

File: api/v1alpha1/packagebundlecontroller_webhook.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+	resp := &admission.Response{
+		AdmissionResponse: admissionv1.AdmissionResponse{Allowed: found},
+	}
+
+	if !found {
+		msg := fmt.Sprintf("package bundle not found with name: %q", pbc.Spec.ActiveBundle)
+		resp.AdmissionResponse.Result = &metav1.Status{
+			Status:  metav1.StatusFailure,
+			Code:    http.StatusUnprocessableEntity,
```
Comment (lewisdiamond) [inline api/v1alpha1/packagebundlecontroller_webhook.go:89]:
```suggestion
			Code:    http.StatusNotFound,
```

File: api/v1alpha1/packagebundlecontroller_webhook.go [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+			Message: msg,
+			Reason:  metav1.StatusReasonInvalid,
```
Comment (lewisdiamond) [inline api/v1alpha1/packagebundlecontroller_webhook.go:91]:
```suggestion
			Reason:  metav1.StatusReasonNotFound,
```

File: api/v1alpha1/packagebundlecontroller_webhook_test.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+	t.Run("handles list errors", func(t *testing.T) {
+		pbc := &PackageBundleController{
+			Spec: PackageBundleControllerSpec{
+				ActiveBundle: "v1-21-1001",
+			},
+			Status: PackageBundleControllerStatus{},
+		}
+		pbcBytes, err := json.Marshal(pbc)
+		require.NoError(t, err)
+		log.Printf("marshaled %d bytes", len(pbcBytes))
```
Comment (lewisdiamond) [inline api/v1alpha1/packagebundlecontroller_webhook_test.go:86]:
```suggestion
```

File: api/v1alpha1/packagebundlecontroller_webhook_test.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+					},
+					Spec:   PackageBundleSpec{},
+					Status: PackageBundleStatus{},
+				},
+			},
+		}
+		resp, err := v.handleInner(ctx, pbc, bundles)
+		if assert.NoError(t, err) {
+			assert.False(t, resp.AdmissionResponse.Allowed)
+			assert.Equal(t, metav1.StatusFailure, resp.AdmissionResponse.Result.Status)
```
Comment (lewisdiamond) [inline api/v1alpha1/packagebundlecontroller_webhook_test.go:129]:
```suggestion
			assert.Equal(t, metav1.StatusReasonNotFound, resp.AdmissionResponse.Result.Status)
```

File: main.go [hunk 7]
```diff
@@ -94,6 +94,10 @@ func main() {
 			setupLog.Error(err, "unable to create webhook", "webhook", "PackageBundle")
 			os.Exit(1)
 		}
+		if err := api.InitActiveBundleValidator(mgr); err != nil {
```
Comment (lewisdiamond) [inline main.go:97]: Wrap this in the above `if` to disable when running locally as this causes issues due to certificates. Unless you found a workaround?

File: PROJECT [hunk 8]
```diff
@@ -41,7 +44,6 @@ resources:
   path: github.com/aws/eks-anywhere-packages/api/v1alpha1
   version: v1alpha1
   webhooks:
-    defaulting: true
```
[the remark this answers is not in the document]
Comment (ewollesen) [reply PROJECT:44]: > Have you considered keeping the defaulting webhook and setting the latest bundle as active if none is provided?

I guess I could, but that might not buy us anything, as it's possible that there aren't any bundles when the controller is created for it to default to. I guess we could try to query the repo, but that feels yucky to me to do in a webhook. In theory, at creation, things are created in the correct order, with the correct names, so defaulting probably doesn't provide much value.

If you feel strong about it, I'll add it.

File: api/v1alpha1/packagebundlecontroller_webhook_test.go (hunk 6 above)
Comment (ewollesen) [reply api/v1alpha1/packagebundlecontroller_webhook_test.go:129]: This isn't quite what we'd want.

The Status field and the Reason field are different things.

I can add your suggest, but my original code is still valid.

File: main.go (hunk 7 above)
Comment (lewisdiamond) [reply main.go:97]: Ah it's the tabs issue.

**Review by lewisdiamond — APPROVED** [review]

**Pull request merged**